iree/test/e2e/regression/matmul_add.mlir

```diff
@@ -1,26 +1,271 @@
+// RUN: [[ $IREE_VULKAN_DISABLE == 1 ]] || (iree-run-mlir -export-all -iree-hal-target-backends=vulkan-spirv %s | IreeFileCheck %s)
 
-func @matmul_add() -> tensor<2x4xf32> {
-  %0 = iree.unfoldable_constant dense<[
-    [1.0,  2.0,  3.0,  4.0,  5.0,  6.0,  7.0,  8.0],
-    [9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0]]> : tensor<2x8xf32>
-  %1 = iree.unfoldable_constant dense<[
-    [ 1.0,  2.0,  3.0,  4.0],
-    [ 5.0,  6.0,  7.0,  8.0],
-    [ 9.0, 10.0, 11.0, 12.0],
-    [13.0, 14.0, 15.0, 16.0],
-    [17.0, 18.0, 19.0, 20.0],
-    [21.0, 22.0, 23.0, 24.0],
-    [25.0, 26.0, 27.0, 28.0],
-    [29.0, 30.0, 31.0, 32.0]]> : tensor<8x4xf32>
-  %2 = iree.unfoldable_constant dense<[
-    [1.0, 2.0, 3.0, 4.0],
-    [5.0, 6.0, 7.0, 8.0]]> : tensor<2x4xf32>
-  %3 = "mhlo.dot"(%0, %1) : (tensor<2x8xf32>, tensor<8x4xf32>) -> tensor<2x4xf32>
-  %4 = mhlo.add %3, %2 : tensor<2x4xf32>
-  return %4 : tensor<2x4xf32>
+// func @matmul_add() -> tensor<2x4xf32> {
+//   %0 = iree.unfoldable_constant dense<[
+//     [1.0,  2.0,  3.0,  4.0,  5.0,  6.0,  7.0,  8.0],
+//     [9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0]]> : tensor<2x8xf32>
+//   %1 = iree.unfoldable_constant dense<[
+//     [ 1.0,  2.0,  3.0,  4.0],
+//     [ 5.0,  6.0,  7.0,  8.0],
+//     [ 9.0, 10.0, 11.0, 12.0],
+//     [13.0, 14.0, 15.0, 16.0],
+//     [17.0, 18.0, 19.0, 20.0],
+//     [21.0, 22.0, 23.0, 24.0],
+//     [25.0, 26.0, 27.0, 28.0],
+//     [29.0, 30.0, 31.0, 32.0]]> : tensor<8x4xf32>
+//   %2 = iree.unfoldable_constant dense<[
+//     [1.0, 2.0, 3.0, 4.0],
+//     [5.0, 6.0, 7.0, 8.0]]> : tensor<2x4xf32>
+//   %workload = constant 8 : index
+//   %3 = flow.dispatch.region[%workload: index]
+//     (%arg0 = %0 : tensor<2x8xf32>, %arg1 = %1 : tensor<8x4xf32>,
+//      %arg2 = %2 : tensor<2x4xf32>) -> tensor<2x4xf32> {
+//     %4 = "mhlo.dot"(%arg0, %arg1) :
+//       (tensor<2x8xf32>, tensor<8x4xf32>) -> tensor<2x4xf32>
+//     %5 = mhlo.add %4, %arg2 : tensor<2x4xf32>
+//     flow.return %5 : tensor<2x4xf32>
+//   }
+//   return %3 : tensor<2x4xf32>
+// }
+
+// // CHECK: EXEC @matmul_add
+// // CHECK: 2x4xf32=[709 746 783 820][1673 1774 1875 1976]
+
+func @matmul_add_large() -> tensor<16x48xf32> {
+  %0 = iree.unfoldable_constant dense<2.0> : tensor<16x32xf32>
+  %1 = iree.unfoldable_constant dense<3.0> : tensor<32x48xf32>
+  %2 = iree.unfoldable_constant dense<4.0> : tensor<16x48xf32>
+  %workload = constant 768 : index
+  %3 = flow.dispatch.region[%workload: index]
+    (%arg0 = %0 : tensor<16x32xf32>, %arg1 = %1 : tensor<32x48xf32>,
+     %arg2 = %2 : tensor<16x48xf32>) -> tensor<16x48xf32> {
+    %4 = "mhlo.dot"(%arg0, %arg1) :
+      (tensor<16x32xf32>, tensor<32x48xf32>) -> tensor<16x48xf32>
+    %5 = mhlo.add %4, %arg2 : tensor<16x48xf32>
+    flow.return %5 : tensor<16x48xf32>
+  }
+  return %3 : tensor<16x48xf32>
 }
+//      CHECK: EXEC @matmul_add_large
+//      CHECK: 16x48xf32=
+// CHECK-SAME: [196 196 196 196 196 196 196 196 196 196 196 196
```

Comment:
feels suspicious - is there something more useful we can use to ensure this is producing the correct results without transposition/etc?

Reply:
The answer is right. Each element is (2.0 * 3.0) * 32 + 4.0 = 196

Reply:
Tried to do something more robust (though I think what is there is fine). Tried to take the output of the following numpy program

```
import numpy as np

np.set_printoptions(threshold=8192, precision=6)
a = np.arange(0.0, 1.0, 1.0/(16*32), np.float32).reshape((16, 32))
b = np.arange(0.0, 1.0, 1.0/(32*48), np.float32).reshape((32, 48))
c = np.arange(0.0, 1.0, 1.0/(16*48), np.float32).reshape((16, 48))
d = np.matmul(a, b) + c
print (repr(a))
print (repr(b))
print (repr(c))
print (d)
```

But the `iree-run-mlir` doesnt have the check dialect. So testing the output becomes almost impossible

Reply:
Moved the test to `structural/...` to  use the `check.*` dialect for correctness.